File: Algorithm/BruteForce/전력망을 둘로 나누기.swift
import Foundation

func solution(_ n:Int, _ wires:[[Int]]) -> Int {
    var answer: Int = n
    var adj:[[Int]] = Array(repeating: Array(repeating: 0, count: n + 1), count: n + 1)
    
    for wire in wires {
        adj[wire[0]][wire[1]] = 1
        adj[wire[1]][wire[0]] = 1
    }
    
    for wire in wires {
        adj[wire[0]][wire[1]] = 0
        adj[wire[1]][wire[0]] = 0
        
        let ret1 = BFS(n, adj, wire[0])
        let ret2 = BFS(n, adj, wire[1])
        
        answer = min(answer, abs(ret1 - ret2))
        adj[wire[0]][wire[1]] = 1
        adj[wire[1]][wire[0]] = 1
    }
    
    return answer
}

func BFS(_ n:Int, _ adj:[[Int]], _ start: Int) -> Int {
    var res: Int = 0
    var visited: [Bool] = Array(repeating: false, count: n + 1)
    var que:[Int] = []
    
    visited[start] = true
    que.append(start)
    
    while !que.isEmpty {
        let cur = que.removeFirst()
        res += 1
        
        for i in 1...n {
            if visited[i] || adj[cur][i] == 0{
                continue
            }
            
            visited[i] = true
            que.append(i)
        }
    }
    
    return res
}

File: Algorithm/BruteForce/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<vector<int>> adj;

int BFS(int n, int start);
int abs(int n);
int min(int a, int b);

int solution(int n, vector<vector<int>> wires) {
    int answer = 201;
    adj = vector<vector<int>>(n + 1, vector<int>(n+1, 0));
    
    for (const vector<int> wire: wires) {
        int a = wire[0], b = wire[1];
        adj[a][b] = 1;
        adj[b][a] = 1;
    }
    
    for (const vector<int> wire: wires) {
        adj[wire[0]][wire[1]] = 0;
        adj[wire[1]][wire[0]] = 0;
        
        int a = BFS(n, wire[0]);
        int b = BFS(n, wire[1]);
        
        answer = min(answer, abs(a-b));
        
        adj[wire[0]][wire[1]] = 1;
        adj[wire[1]][wire[0]] = 1;
    }
    
    return answer;
}

int BFS(int n, int start) {
    int res = 1;
    vector<int> visited(n + 1, false);
    queue<int> que;
    
    que.push(start);
    visited[start] = true;
    
    while (!que.empty()) {
        int cur = que.front();
        que.pop();
        
        for (int i = 1; i <= n; ++i) {
            if (!adj[cur][i] || visited[i])
                continue;
            
            visited[i] = true;
            res++;
            que.push(i);
        }
    }
    
    return res;
}

int abs(int n) {
    return n < 0 ? -n : n;
}

int min(int a, int b) {
    return a < b ? a : b;
}